### DISCUSSION TITLE
About mixing FSM with behavior tree

### DISCUSSION THREAD

**Discussion by MadFlyFish** [kagenash1/godot-behavior-tree#19] [General]
Interrupting flow #6 
About the Interrupting flow issue you guys talked about before, I wonder if there is any way to combine the advantage of  FSM(clear state switch) and behavior tree(low coupling, extensibility)
:)
![image](https://user-images.githubusercontent.com/36298330/116177875-2d551780-a747-11eb-8da9-fcda277e9402.png)

**Comment by kagenash1** [comment]
Technically yes, there should be no problem. But you need to sync the start/abortion of trees with the state switch.

I do something similar by using a state machine to update a "state" variable in the Blackboard, and then only have one tree with parallel nodes which correnspond to the different states.

In this way you don't worry about syncing problem and the tree becomes responsive to the state change.

**Reply by MadFlyFish** [reply]
Yep, I had thought of some kind of solution just like what you described, but I haven't tested it yep.
However, it is quite confusing how to abort tree branches immediately and completely, especially when coroutines are frequently used, or some leaf takes a long time to complete execution.

**Reply by kagenash1** [reply]
That can easily be done in the new version .

You can have a BTSelector, this selector has several BTRepeatUntil with a "Failure" flag set and a BTConditional inside.
The BTConditional checks if you are still in the desired state (which is stored in the blackboard).
You can update the state from the outside and the BTConditional will respond to that change and fail, allowing you to abort the current branch and go to another branch for the new state.

Plus, now you can even abort the entire tree after a node reached completion just by toggling a flag in the inspector.

However an action should always be designed so that it's run to completion. So you should think of it in "steps", for example shooting a single bullet, rather than shooting until the target is dead, or doing a full swing of the sword before going back to "chase" mode.
It's important that your actions are short enough to let you re-evaluate often. 

This is the case even with a state machine: you exit a state only after completing the current "step".

If you wanna allow actions to be mixed together and switch state immediately you can use a BTParallel instead of a selector. That way you will immediately execute the branch for the new state. But that can be risky cause the previous state still needs to complete, so you should really just do it for behaviors that can blend in easily.

**Comment by MadFlyFish** [comment]
Yep I  see what you mean. 
But these lines are not very clear to me. Please excuse me for my poor English :)
I would be grateful if you share your screenshot of the structures described below


![image](https://user-images.githubusercontent.com/36298330/116339904-d7e53d00-a810-11eb-8421-7854b7bb9ae9.png)


By the way, I'm in similar discussions with the author of XSM.
https://gitlab.com/atnb/xsm/-/issues/4

**Comment by MadFlyFish** [comment]
I try it and the state switching works well, just like FSM does. (Please point out If I misunderstand )
And in this case it seems to work properly without the BTRepeatUntil. So why is BTRepeatUntil necessary here?

In addition to these, I still don't have a proven way to end some state immediately. Maybe I should make these actions as short as possible. LOL

![image](https://user-images.githubusercontent.com/36298330/116361187-489d5100-a833-11eb-91c8-b870f6cd052f.png)
![image](https://user-images.githubusercontent.com/36298330/116361228-55ba4000-a833-11eb-902e-30cb50e35b2c.png)
![image](https://user-images.githubusercontent.com/36298330/116361272-5fdc3e80-a833-11eb-8ab3-05ae18f2259d.png)

**Reply by kagenash1** [reply]
The repeater is needed if you want to remain inside a tree branch, like a state machine. 
Without it you wil still tick the whole tree at every frame.

Your states are interrupted when the conditional fails, but the action still needs to complete in order to re evaluate the condition at the next tick, just like a state machine does.

You can't interrupt it during its execution because you can't interrupt a function during its execution (unless you do it manually within the function itself) it's not the plugin's fault.

Again, your actions should be thought in "steps", if you have a long action inside your leaf script it is a design problem.

If you have a behavior that you want to interrupt when the state is switched, then you should split in multiple leaves so that you can re evaluate the condition between them and interrupt if necessary.

**Comment by MadFlyFish** [comment]
Thanks. This gives me a deeper understanding about how behavior tree should be designed.
I will keep testing the addon.  :)

**Reply by kagenash1** [reply]
Anyway, I had an idea to allow re evaluation of conditionals even before completion. For example, if you have:

BTConditional -> BTSequence -> [BTLeaf1, BTLeaf2, ...] 

You wanna re evaluate the condition at each step of the sequence. This will probably result in a "conditional" sequence that allows you to interrupt a sequence before completion if a condition changes.

EDIT: Actually this can be done in a more general way, but will most likely be a bigger update. So for now this will remain a limitation! It will address the interrupting flow issue.
